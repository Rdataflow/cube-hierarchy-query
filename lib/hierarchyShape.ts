import clownface, { GraphPointer } from 'clownface'
import { rdf, sh, dashSparql } from '@tpluscode/rdf-ns-builders'
import { meta } from '@zazuko/vocabulary-extras/builders'
import { isGraphPointer, isNamedNode } from 'is-graph-pointer'
import { s2q } from '@hydrofoil/shape-to-query'

export function fromHierarchy(hierarchy: GraphPointer) {
  const clone = clownface({
    dataset: hierarchy.dataset.match(),
    term: hierarchy.term,
  })

  const rootShape = clone.blankNode()

  rootShape
    .addOut(rdf.type, sh.NodeShape)
    .addOut(sh.targetNode, hierarchy.out(meta.hierarchyRoot))
    .addOut(sh.rule, rule => {
      rule.addOut(rdf.type, s2q.SPORule)
    })

  let currentLevel = hierarchy.out(meta.nextInHierarchy)
  let currentLevelShape: GraphPointer = rootShape

  while (currentLevel) {
    const path = currentLevel.out(sh.path)
    if (!isGraphPointer(path)) {
      break
    }

    let nextLevelShape : GraphPointer

    currentLevelShape
      .addOut(sh.property, constraint => {
        constraint.addOut(sh.path, path)
          .addOut(sh.node, nodeConstraint => {
            nodeConstraint.addOut(sh.rule, rule => {
              rule.addOut(rdf.type, s2q.SPORule)
            })

            nextLevelShape = nodeConstraint
          })

        const targetClass = currentLevel.out(sh.targetClass).term
        if (targetClass) {
          constraint.addOut(sh.class, targetClass)
        }

        if (isNamedNode(path)) {
          currentLevelShape.out(sh.rule)
            .addOut(s2q.predicateFilter, propertyFilter(path))
        }
      })

    currentLevel = currentLevel.out(meta.nextInHierarchy)
    currentLevelShape = nextLevelShape
  }

  return rootShape
}

function propertyFilter(path: GraphPointer) {
  return (ptr: GraphPointer) => {
    ptr.addList(dashSparql.not, [
      ptr.blankNode().addList(dashSparql.eq, [
        sh.this,
        path,
      ]),
    ])
  }
}
