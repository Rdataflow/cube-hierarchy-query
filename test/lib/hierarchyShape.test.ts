import { meta } from '@zazuko/vocabulary-extras/builders'
import { schema, sh } from '@tpluscode/rdf-ns-builders'
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import { fromHierarchy } from '../../lib/hierarchyShape.js'
import { ex, parse, serialize } from '../support.js'

describe('lib/hierarchyShape', () => {
  chai.use(jestSnapshotPlugin())

  describe('fromHierarchy', () => {
    it('excludes property from SPO rule', async () => {
      // given
      const hierarchy = await parse`
        <>
          ${meta.hierarchyRoot} <Europe> ;
          ${meta.nextInHierarchy} <firstLevel> ;
        .
        <firstLevel> 
          ${sh.path} ${schema.containsPlace} ;
          ${meta.nextInHierarchy} <secondLevel> .
        <secondLevel> 
          ${sh.path} ${schema.containsPlace} .
      `

      // when
      const shape = fromHierarchy(hierarchy.namedNode(ex()))

      // then
      expect(
        await serialize(shape),
      ).toMatchSnapshot()
    })

    it('does not add excludes when properties are inverse', async () => {
      // given
      const hierarchy = await parse`
        <>
          ${meta.hierarchyRoot} <Europe> ;
          ${meta.nextInHierarchy} <firstLevel> ;
        .
        <firstLevel> 
          ${sh.path} [ ${sh.inversePath} ${schema.containedInPlace} ] ;
          ${meta.nextInHierarchy} <secondLevel> .
        <secondLevel> 
          ${sh.path} [ ${sh.inversePath} ${schema.containedInPlace} ] ; .
      `

      // when
      const shape = fromHierarchy(hierarchy.namedNode(ex()))

      // then
      expect(
        await serialize(shape),
      ).toMatchSnapshot()
    })
  })
})
