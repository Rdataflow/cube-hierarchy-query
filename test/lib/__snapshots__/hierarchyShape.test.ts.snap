// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/hierarchyShape fromHierarchy does not add excludes when properties are inverse 1`] = `
"@prefix cube: <https://cube.link/> .
@prefix meta: <https://cube.link/meta/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix schema: <http://schema.org/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix s2q: <https://hypermedia.app/shape-to-query#> .

_:b17 a sh:NodeShape ;
	sh:targetNode <http://example.com/Europe> ;
	sh:rule [
		rdf:type s2q:SPORule ;
	] ;
	sh:property [
		sh:path [
			sh:inversePath schema:containedInPlace ;
		] ;
		sh:node [
			sh:rule [
				rdf:type s2q:SPORule ;
			] ;
			sh:property [
				sh:path [
					sh:inversePath schema:containedInPlace ;
				] ;
			] ;
		] ;
	] .

"
`;

exports[`lib/hierarchyShape fromHierarchy excludes property from SPO rule 1`] = `
"@prefix cube: <https://cube.link/> .
@prefix meta: <https://cube.link/meta/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix schema: <http://schema.org/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix s2q: <https://hypermedia.app/shape-to-query#> .

_:b1 a sh:NodeShape ;
	sh:targetNode <http://example.com/Europe> ;
	sh:rule [
		rdf:type s2q:SPORule ;
		s2q:predicateFilter [
			<http://datashapes.org/sparql#not> (
				[
					<http://datashapes.org/sparql#eq> (
						sh:this
						schema:containsPlace
					) ;
				]
			) ;
		] ;
	] ;
	sh:property [
		sh:path schema:containsPlace ;
		sh:node [
			sh:rule [
				rdf:type s2q:SPORule ;
				s2q:predicateFilter [
					<http://datashapes.org/sparql#not> (
						[
							<http://datashapes.org/sparql#eq> (
								sh:this
								schema:containsPlace
							) ;
						]
					) ;
				] ;
			] ;
			sh:property [
				sh:path schema:containsPlace ;
			] ;
		] ;
	] .

"
`;
